### PR TITLE
Complete the wire-serialization layer

### PR DESCRIPTION
**Pull request by ProofOfKeags**
Completes the wire-serialization layer: every byte of a Bitcoin block now parses through a verified codec, anchored to real mainnet data, with the test and CI infrastructure to keep it that way.

## The serialization stack

- **Codec discipline** (`Serialize/Codec.lean`): the `Codec` typeclass — encoder, prefix-consuming decoder over `List UInt8`, and the two laws (round-trip, canonicality) read as the encode/decode adjunction. Product composition and `Codec.ofEquiv` transport let composite structures inherit correctness from their fields; one little-endian `BitVec (8 * n)` primitive defines byte order once for all fixed-width integers and the 256-bit hash type.
- **CompactSize** rebuilt on those primitives and moved beside them, gaining the first-byte-is-zero-only-for-zero lemma the SegWit marker dispatch needs; **`CountedList`** captures every CompactSize-counted wire field (scripts, vectors, witness stacks, block txs) as one lifted codec.
- **`Script`**: programs get their own type while the text stays raw — a protocol fact (tokenization belongs to execution and may fail; consensus never requires script fields to tokenize), not a simplification.

## The data model and its codecs

The single `Block.Types` module becomes one leaf per structure (`OutPoint`, `TxIn`, `TxOut`, `WitnessStack`, `SegwitInput`, `TxBody`, `Tx`, `BlockHeader`, `Block`), each with its codec. The transaction model is rebuilt around the wire format's own structure: an inductive with `legacy` and `segwit` constructors over a shared witness-free `TxBody` (txid preimage = legacy serialization = witness-stripped view); one-witness-per-input arity is structural; the legacy constructor carries a non-empty-inputs proof because `0x00` is the reserved SegWit marker. Plain products get codecs by composition with no hand-written proofs; the whole-transaction codec is the one hand-written case (BIP144 witness regrouping) and proves round-trip and canonicality including the marker dispatch.

## SHA-256

A concrete, computable FIPS 180-4 `sha256`/`sha256d`, validated against published test vectors. Collision resistance is deliberately not stated — provably false for a concrete hash by pigeonhole — so it can only ever appear as a hypothesis over an abstract hash, never an axiom over this function.

## Tests and CI

- **Golden vectors** (`#guard` at build time): real mainnet bytes through the verified decoders — the first Bitcoin payment, the SegWit activation coinbase, the first SegWit spend, the genesis block, and block 170 (whose embedded payment must re-encode to the standalone vector), plus CompactSize boundary values and SHA-256 known-answer tests.
- **Axiom audit**: the build fails if any registered theorem depends on `sorryAx` or an axiom beyond the standard three (`bv_decide` LRAT certificates recognized and allowed) — turning "it builds" into "it is proved".
- **`lake test`**: fetches block 481824 (the SegWit activation block, 1866 transactions mixing both serialization eras) by hash on first run, caches it gitignored, verifies the header double-SHA-256s to the requested hash, spot-checks it, and re-encodes all 989,323 bytes. Public chain data, so never committed.
- CI runs build, test, and lint; a green badge now means proved, linter-clean, and matching mainnet bytes.

CLAUDE.md / CONTRIBUTING.md record the build loop, the codec discipline, and the leaf-landing checklist.

Every commit in the series builds, so `git bisect` works across the branch. The `merkle-commitment` branch (#7) is stacked on this one.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .claude/skills/proof-leaf/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+# Landing a proof leaf
+
+A "proof leaf" is this repo's unit of progress: one module that builds
+cleanly, states its checked claims, and is wired into the build, the audit,
+and the README. Follow this checklist end to end — a leaf is not landed until
+every step is done.
+
+## 1. The module
+
+- One concern per file, under the matching directory (`Serialize/`,
```
Comment (ProofOfKeags) [inline .claude/skills/proof-leaf/SKILL.md:15]: I want to state a better model of "concern". One of the things we ran into was that Codec appeared under  separate module to be a bucket of typeclass instances where the types and the class were both defined elsewhere. I'm of the opinion typeclass instances should live hear the types, especially for any typeclasses we define in this module. Obviously there may be an exception if we define instances of our typeclass for types in Core or something deeper, but this is how we should think about that.

Per that, I think that means that we should split modules by one major type. We should avoid defining more than one type per module. The exception here is if we have record types we use for different arms of a sum type then it's all generally ok to live in one module if those arm subtypes don't show up outside the module in any function signatures.

File: .claude/skills/proof-leaf/SKILL.md
Comment (ProofOfKeags) [inline .claude/skills/proof-leaf/SKILL.md:1]: I think we need a way also to audit the codebase against this spec for module discipline.

File: BtcVerified/Crypto/Sha256.lean
Comment (ProofOfKeags) [inline BtcVerified/Crypto/Sha256.lean:1]: I hadn't realized we were going to literally implement sha256 but I suppose that might be required. Is this a thing at all offered by CSLib? If not, are there any cryptography libraries we should think about importing? Otherwise I'd like to extensively test/verify this implementation to the extent possible.

File: BtcVerified/Ext/List.lean [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  regions and zip it back, but nothing here mentions transactions.
+-/
+
+namespace List
+
+variable {α β γ : Type*}
+
+/-- Recombine a list's two projections with their constructor to recover the
+list, when the constructor and projections round-trip. -/
+theorem zipWith_map_proj {f : γ → α} {g : γ → β} {mk : α → β → γ}
```
Comment (ProofOfKeags) [inline BtcVerified/Ext/List.lean:20]: I always struggle to remember theorem names like this. Is there a way to imagine what the name would be knowing only the type?

File: .claude/skills/proof-leaf/SKILL.md (hunk 1 above)
Comment (ProofOfKeags) [reply .claude/skills/proof-leaf/SKILL.md:15]: Agreed — let's make this the explicit rule, with the edges pinned down:

- **One type per module**; that type's instances (especially of classes we define, like `Codec`) live in its module, never a collector.
- **Exception A**: instances of *our* class over a Core/dependency type live with the *class*, since we can't touch the type's module — that's why `instCodecUInt8…64`/`instCodecBitVec256` sit in `Serialize/Codec.lean`.
- **Exception B**: record arms of a sum type may share a module *only if* those arm types never appear in a signature outside it. This is exactly why `SegwitInput` is its own module rather than an arm-record: it shows up in `Tx`'s constructor and the codec signatures.

For the record, this branch already complies — every new file defines one type and carries its own codec. The one violator is the pre-existing `BitVM/BitCommitment.lean` (5 types, where `Opening`/`Collision` do appear in signatures), which is exactly the kind of thing the audit (next thread) should surface for a deliberate decision.

Holding the SKILL.md/CLAUDE.md wording change for the follow-up branch per your call to keep #8 as-is.

File: .claude/skills/proof-leaf/SKILL.md
Comment (ProofOfKeags) [reply .claude/skills/proof-leaf/SKILL.md:1]: Let's build this as a **Lean environment-introspection audit** rather than a syntactic grep checker — correct rather than approximate, and it reuses the `import-graph` dependency we already have.

Shape: an executable (sibling to the `tests` driver) that imports the whole library and walks `Lean.Environment` to enforce, per the rule in the thread above:
1. each module declares ≤1 top-level type (`structure`/`inductive`), with an allowlist for sum-type arm-record clusters;
2. every `instance` of a btc-verified-defined class targets a type declared in the same module, *unless* the target is an allowlisted Core/dependency type.

It's the same posture as the existing `#assert_axioms` audit — machine-checked discipline, not convention — just whole-codebase instead of per-declaration. Deferring the build to the follow-up branch; capturing the decision here.

File: BtcVerified/Crypto/Sha256.lean
Comment (ProofOfKeags) [reply BtcVerified/Crypto/Sha256.lean:1]: Checked the ecosystem: **CSLib does not offer SHA-256** — it's CS *foundations* (semantics, automata, concurrency), the Mathlib-for-CS, not a primitives library. And there's **no mature verified SHA-256 anywhere in Lean** right now; closest prior art is SHA-3 *experiments* ([gdncc/Cryptography](https://github.com/gdncc/Cryptography), [Doussot's SHA-3 paper](https://eprint.iacr.org/2024/1880.pdf)) and AWS's s2n-bignum port (bignum arithmetic, not hashes). So implementing it ourselves is the right call.

Agreed direction on verification: **expand KATs + lean on real-block anchoring.**
- Add the NIST CAVP short/long-message vectors and more real Bitcoin block hashes to the `#guard` suite.
- Treat the real-block anchoring as the practical proof — the 481824 fixture already verifies the header `sha256d` *and* the merkle commitment over all 1866 txids, which is strong evidence the function matches the network's.
- File full formal correctness (proving the impl realizes the FIPS 180-4 relation) as a someday-leaf — it's research-scale and blocks nothing.

One smell to log for that someday-leaf: the impl uses `Array` + `get!` panic-indexing inside `Id.run do`. Fine for *computing* (KATs pass), but proving anything about it would want bounds-safe indexing first. Deferring all of this to the follow-up branch.

File: BtcVerified/Ext/List.lean (hunk 2 above)
Comment (ProofOfKeags) [reply BtcVerified/Ext/List.lean:20]: Yes — that's exactly Mathlib's naming discipline: **a theorem name is a transcription of its statement**, head symbols of the LHS read outside-in, joined by `_`, with relation words (`eq`/`le`/`lt`), connectives (`of`/`iff`), and direction tags (`left`/`right`). `zipWith_map_proj` is a loose rendering; a stricter one would be closer to `zipWith_map_map_left_right`. The whole point of the convention is your wish — names become *derivable* rather than memorized.

And when derivation fails, three find-by-**type** tools (we already depend on the last):
- `exact?` / `apply?` / `rw?` — search by the goal's type;
- **loogle** — search by term pattern/shape;
- `#leansearch` (LeanSearchClient, already a dep) — natural-language search.

So the workflow is: don't memorize, query. Concrete fixup for the follow-up branch: tighten our `Ext/` names toward the mechanical-transcription convention so they're reconstructible the way you want.

File: .claude/skills/proof-leaf/SKILL.md (hunk 1 above)
Comment (ProofOfKeags) [reply .claude/skills/proof-leaf/SKILL.md:15]: add a commit splitting the bitcommitment file into types. Naively is fine. The BitVM segment is on ice for now.

File: .claude/skills/proof-leaf/SKILL.md
Comment (ProofOfKeags) [reply .claude/skills/proof-leaf/SKILL.md:1]: We need a policy where if we defer work we know we need to do, deferral is not considered "complete" until the work item has been generated as an issue.

File: BtcVerified/Crypto/Sha256.lean
Comment (ProofOfKeags) [reply BtcVerified/Crypto/Sha256.lean:1]: same policy here regarding deferral

File: BtcVerified/Ext/List.lean (hunk 2 above)
Comment (ProofOfKeags) [reply BtcVerified/Ext/List.lean:20]: I'm assuming agents know how to use the find-by-type tools. What reference should I use to make them available to me?

File: .claude/skills/proof-leaf/SKILL.md (hunk 1 above)
Comment (ProofOfKeags) [reply .claude/skills/proof-leaf/SKILL.md:15]: Queued. One wrinkle worth naming: `BitCommitment.lean` lives on **master** (unchanged through #8 and #7), while the doc files that should carry the codified rule (`CLAUDE.md`/`SKILL.md`) only exist on **#8**. So the split and the rule-codification want a shared base — I'll fold both into one "module discipline" PR (placement question on the audit thread). The split itself is the naive five-module break: `Bit`, `Opening` (+ the `commit`/`Verifies` machinery), `ValidOpening`, `Equivocation`, `Collision` — the `Prop` predicates (`Binding`/`NonEquivocation`/`CollisionResistance`) aren't `structure`/`inductive`, so they sit with the type they're about without breaking the one-type rule.

File: .claude/skills/proof-leaf/SKILL.md
Comment (ProofOfKeags) [reply .claude/skills/proof-leaf/SKILL.md:1]: Audit filed as #9 (the introspection-executable shape, wired into CI like `lake test`/`lake lint`).

On the deferral policy — adopted, and applied retroactively: every deferred item from this review now has an issue (#9 audit, #10 SHA KATs, #11 SHA formal-correctness, #12 Ext naming). I'll also codify the policy itself in `CLAUDE.md` so future sessions honour it: *a deferral isn't "done" until the work is a tracked issue.* That codification rides in the same module-discipline PR (placement below).

File: BtcVerified/Crypto/Sha256.lean
Comment (ProofOfKeags) [reply BtcVerified/Crypto/Sha256.lean:1]: Filed: **#10** (expand KATs — NIST CAVP short/long + more real block hashes) and **#11** (someday-leaf: prove the impl realizes the FIPS 180-4 relation, with the bounds-safe-indexing refactor as its first step). Practical assurance stays "KATs + real-block anchoring"; #11 is tracked but blocks nothing.

File: BtcVerified/Ext/List.lean (hunk 2 above)
Comment (ProofOfKeags) [reply BtcVerified/Ext/List.lean:20]: References, in order of how I'd reach for them:

- **`exact?` / `apply?` / `rw?`** — built into Lean+Mathlib, zero setup: put your cursor on the goal and they search by the goal's *type*. This is the day-to-day "find the lemma from the type" tool.
- **loogle** — search by term *pattern/shape* (e.g. `List.zipWith (_ + _)`): the web UI at https://loogle.lean-lang.org/, and the `#loogle` command in-editor (Mathlib's `Mathlib.Tactic.Loogle`).
- **`#leansearch`** — natural-language queries (leansearch.net) via LeanSearchClient, which we already depend on — usable in any file in this repo today.
- The convention these names follow: the Mathlib naming guide, https://leanprover-community.github.io/contribute/naming.html — that's the reference for *reconstructing* a name from a statement.

Happy to add a short "Finding lemmas by type" section to `CONTRIBUTING.md` pointing at these — say the word and it goes in the module-discipline PR.

File: .claude/skills/proof-leaf/SKILL.md (hunk 1 above)
Comment (ProofOfKeags) [reply .claude/skills/proof-leaf/SKILL.md:15]: Done in #13: the naive five-module split (`BitVM/BitCommitment/` under an umbrella facade, namespace and audit registrations unchanged) plus the rule codified in `CLAUDE.md`/`CONTRIBUTING.md`/the proof-leaf skill. Stacked on this branch, merges after it.

File: .claude/skills/proof-leaf/SKILL.md
Comment (ProofOfKeags) [reply .claude/skills/proof-leaf/SKILL.md:1]: Policy codified in `CLAUDE.md` (and `CONTRIBUTING.md`) in #13: *a deferral isn't done until it's a tracked issue.* The audit itself is #9.

File: BtcVerified/Ext/List.lean (hunk 2 above)
Comment (ProofOfKeags) [reply BtcVerified/Ext/List.lean:20]: Added the "Finding lemmas by type" section to `CONTRIBUTING.md` in #13 — `exact?`/`apply?`/`rw?`, loogle, `#leansearch`, and the naming guide, so it's a durable reference, not just this thread.
